### ISSUE TITLE
Not bound composite part stops reading of successive parts

### ISSUE THREAD

**Issue by sttz** [Unity-Technologies/InputSystem#875]
If a part in a composite is not bound to anything, it will prevent the parts that follow it from being read.

I.e. create a 2D composite bound to the four arrow keys and then unbind Left. Notice how Up and Down still work but Right doesn't.

If I don't need the parts, I can simply delete them, but I need to unbind parts dynamically at runtime to avoid conflicts between actions after rebinding. If I unbind a part of a composite, part of the composite then stops working.

The problem is how [InputBindingResolver.AddActionMap](https://github.com/Unity-Technologies/InputSystem/blob/stable/Packages/com.unity.inputsystem/InputSystem/Actions/InputBindingResolver.cs#L189) handles bindings with an empty `effectivePath`. It simply skips initializing the binding state, which causes its `isPartOfComposite` to be `false`.

When the composite value is then read in [InputActionState.ReadCompositePartValue](https://github.com/Unity-Technologies/InputSystem/blob/stable/Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs#L1976) during a state change, the method iterates through the parts as long as `isPartOfComposite` is `true`. As soon as it encounters the uninitialized part mentioned above, it stops the iteration and returns `default(TValue)`.

I don't know what a proper fix would be but as a hack, I tried to copy over `isPartOfComposite` in case `effectivePath` is empty in `InputBindingResolver.AddActionMap`, and that fixed the issue:
```cs
// Skip binding if it is disabled (path is empty string).
var path = unresolvedBinding.effectivePath;
if (string.IsNullOrEmpty(path)) {
    bindingState->isPartOfComposite = unresolvedBinding.isPartOfComposite;
    continue;
}
```

**Comment by Rene-Damm** [comment]
Good catch. Thank you for reporting. We'll take a look.

We're in the process of migrating all our bug tracking to the official Unity bug tracking system (which gives us the benefit of properly integrating ourselves with Unity QA procedures) and turn off GitHub issue tracking once done. I've migrated the case over to FogBugz.

FB Ticket: 1189909
Unity Issue Tracker: https://issuetracker.unity3d.com/product/unity/issues/guid/1189909/

Closing the issue here. No worries, doesn't mean it's forgotten :)
